specs/rollup-node.md

```diff
@@ -175,58 +235,77 @@ The following scenarios are assimilated to errors:
-it. This is a duration of approximately 7 days worth of L1 blocks.
+> **TODO** specify genesis block
 
-L1 Ethereum [reaches finality approximately every 12 minutes][l1-finality], so these L2 blocks can safely be considered
-to be final: they will never disappear from the chain's history because of a re-org.
+# Handling L1 Re-Orgs
 
-[l1-finality]: https://www.paradigm.xyz/2021/07/ethereum-reorgs-after-the-merge/
+[l1-reorgs]: #handling-L1-re-orgs
+
+The [previous section on L2 chain derivation][l2-chain-derivation] assumes linear progression of the L1 chain. It is
+also applicable for batch processing, meaning that any given point in time, the canonical L2 chain is given by
+processing the whole L1 chain since the [L2 chain inception][g-inception].
+
+> By itself, the previous section fully specifies the behaviour of the rollup driver. **The current section is
+> non-specificative** but shows how L1 re-orgs can be handled in practice.
+
+In practice, the L1 chain is processed incrementally. However, the L1 chain may occasionally [re-organize][g-reorg],
+meaning the head of the L1 chain changes to a block that is not the child of the previous head but rather another
+descendant of an ancestor of the previous head. In that case, the rollup driver must first search for the common L1
+ancestor, and can re-derive the L2 chain from that L1 block and onwards.
+
+The starting point of the re-derivation is a pair `(refL2, nextRefL1)` where `refL2` refers to the L2 block to build
+upon and `nextRefL1` refers to the next L1 block to derive from (i.e. if `refL2` is derived from L1 block `refL1`,
+`nextRefL1` is the canonicla L1 block at height `l1Number(refL1) + 1`).
+
+In practice, the happy path (no re-org) and the re-org paths are merged. The happy path is simply a special case of the
+re-org path where the starting point of the re-derivation is `(currentL2Head, newL1Block)`.
+
+This re-derivation starting point can be found by applying the following algorithm:
+
+1. (Initialization) Set the initial `refL2` to the head block of the L2 execution engine.
+2. Set `parentL2` to `refL2`'s parent block and `refL1` to the L1 block that `refL2` was derived from.
+3. Fetch `currentL1`, the canonical L1 block at the same height as `refL1`.
+
+- If `currentL1 == refL1`, then `refL2` was built on a canonical L1 block:
+  - Find the next L1 block (it may not exist yet) and return `(refL2, nextRefL1)` as the starting point of the
+    re-derivation.
+    - It is necessary to ensure that no L1 re-org occured during this lookup, i.e. that `nextRefL1.parent == refL1`.
+    - If the next L1 block does not exist yet, there is no re-org, and nothing new to derive, and we can abort the
+        process.
+- Otherwise, if `refL2` is the L2 genesis block, we have re-orged past the genesis block, which is an error that
+    requires a re-genesis of the L2 chain to fix (i.e. creating a new genesis configuration).
```

Comment:
I'm not 100% sure about this, but this my best guess. What happens in practice if we re-org past genesis? Wouldn't it be better to specify that a re-org past genesis is allowed, in which case we just wait for the L2 chain inception L1 block height to restart the whole process?

Reply:
If the L1 reorgs past expected anchor point on L1 then we should just manually redeploy the L2 chain (new L2 genesis) and not build any blocks on top. Wiping the whole L2 chain with a reorg without intervention is pretty severe.

If we start a L2 chain we can probably pick a finalized/confirmed L1 anchor point in the past, to not risk a reorg, at the cost of just a few empty deposit-blocks at the start of the L2 chain.

Reply:
Good point, I'll add a footnote saying just that.